### ISSUE TITLE
io_read request on TCP socket going to Async execution despite not asking for it

### ISSUE THREAD

**Issue by lano1106** [axboe/liburing#351]
Here are the traces:

```
  9233.788 test/622 io_uring:io_uring_submit_sqe(ctx: 0xffff8e9e91b42000, opcode: 22, user_data: 216454257090494477, force_nonblock: 1, flags: 32)
  9233.793 test/622 io_uring:io_uring_poll_arm(ctx: 0xffff8e9e91b42000, opcode: 22, user_data: 216454257090494477, mask: 75, events: 75)
  9233.799 test/622 io_uring:io_uring_cqring_wait(ctx: 0xffff8e9e91b42000, min_events: 1)

  9372.010 test/622 io_uring:io_uring_task_run(ctx: 0xffff8e9e91b42000, opcode: 22, user_data: 216454257090494477)
  9372.026 test/622 io_uring:io_uring_task_run(ctx: 0xffff8e9e91b42000, opcode: 22, user_data: 216454257090494477)

  9372.028 test/622 io_uring:io_uring_queue_async_work(ctx: 0xffff8e9e91b42000, req: 0xffff8e9e84ccbc00, work: 0xffff8e9e84ccbcb8, flags: 24608)
  9372.034 test/622 io_uring:io_uring_complete(ctx: 0xffff8e9e91b42000, user_data: 216454257090494477, res: 1709)
```

My interpretation is that
The file poll function returns that data is available for reading but when io_read() actually tries to get it, the req->file->f_op->read_iter function returns EAGAIN.

Am I interpreting correctly the io_uring traces?

Documenting my issues seems to help to make me realize what is wrong... I might misuse the io_uring API in some way...

I converted nonblocking I/O code to Async with io_uring conversion. With previous nonblocking logic, read() calls were providing larger read buffers than what is possibly available and the caller was absolutely fine and was expecting partial reads result.

The very first thing that I had to change in my io_uring conversion was to remove the O_NONBLOCK flag or else the async read operations were returning EAGAIN immediately.

However, I am not sure if io_uring is settling for partial reads and won't keep reading until the provided buffer is fully filled. My app logs show me that I do receive partial read results but I wonder if the partial reads aren't the cause that makes io_uring punt my requests to io_wq.

Is there a way to indicate to io_uring to not return me EAGAIN but to return immediately the available bytes even if they are less than the requested number?

Update:

For files that are anything else than a regular file, it seems like partial read is what io_read is doing...

```
	} else if (ret <= 0 || ret == io_size || !force_nonblock ||
		   (req->flags & REQ_F_NOWAIT) || !(req->flags & REQ_F_ISREG)) {
		/* read all, failed, already did sync or don't want to retry */
		goto done;
	}
```

thx

**Comment by lano1106** [comment]
events 75 is POLLIN|POLLPRI|POLLERR|POLLRDNORM

concerning the io_uring_poll_arm() trace, I was under the impression that the mask value was the result of the vfs_poll() call. It is not.

If vfs_poll() returns a non-null value, the trace wouldn't be printed.
unless I am mistaken mask will always be equal to events.

I have modified the io_uring_task_run trace to eliminate the possibility that io_poll_rewait() was returning true. It is not.

```
  9287.369 test/625 io_uring:io_uring_task_run(ctx: 0xffff8fbf9a834800, opcode: 22, user_data: 216454257090494477, result: 195)
  9287.386 test/625 io_uring:io_uring_task_run(ctx: 0xffff8fbf9a834800, opcode: 22, user_data: 216454257090494477, result: 195)
```

not sure how result set to 195 is possible when it is ANDed with poll->events that is set to 75...

1. io_uring_task_add() trace is missing. I suspect that trace may be skipped depending on the execution context that this function is run in...
2. idk how io_uring_task_run() can be called twice in a row. This looks like a glitch. __io_async_wake() is removing the wait entry. If io_read() was returning EAGAIN, io_uring_queue_async_work() would be called right away and we wouldn't see 2 consecutive io_uring_task_run() calls. AFAIK REQ_F_POLLED is never cleared...

It is suspicious.

**Comment by lano1106** [comment]
I have augmented the io_uring traces to debug my problem. Some of them are inside io_wq_submit_work():

```
			trace_io_wq_submit_work(req->ctx, req, req->opcode, req->user_data);
			ret = io_issue_sqe(req, 0);
			trace_io_wq_submit_work_completion(ret);
```

I got a surprising result:

```
 31806.742 iou-wrk-629/637 io_uring:io_wq_submit_work_completion(8)
 32158.157 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(1)
 32260.022 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(2)
 32364.793 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(3)
 32844.402 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(8)
 32849.387 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(8)
 32879.902 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(8)
 33006.720 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(0)
 33259.342 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(2)
 33365.219 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(3)
 33394.230 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(3)
```
No idea what those return values are supposed to mean. 0 or -errno is what I would have expected. If this can help to make some sense out of them, the only op executed async is io_read.

Also, I'm a bit surprised to see the io-mgr actually doing some io_wq_submit() calls... Unless it is performed to clean-up at termination...

Also, I have figured out my problem. Who would have thought? It comes from a minor misuse of the API from my program. I submit more than one READ operation on the same fd at the same time.

This had 2 consequences.

1. When data is becoming available, the first req will reap all the available data, the second processed req will fail reading and will go in the io_wq.
2. This is causing a buffer leak. As more and more read seq gets stuck in the io_wq, the provided buffers get depleted.

Takeaways from this experience:
1. If you see io-wrk threads when you do not expect to see any. Investigate, there is something wrong happening
2. It would be nice if io_uring could assist in catching those problems and return an error to the user when it catch them. I predict that I have stumbled into another future io_uring classic problem...

**Comment by lano1106** [comment]
One of the traces modif that I added was to add the req pointer to the following traces:

- io_uring_complete
- io_uring_submit_sqe
- io_uring_poll_arm
- io_uring_task_run

This has been super useful to help me understand what my problem was. Let me know if there would be interest in receiving a patch to get those changes.

**Comment by axboe** [comment]
Definitely interested in that, please do send a patch! It's only when you
start using the traces to debug issues like this that you realize what bits
are missing to make it truly useful.

**Comment by isilence** [comment]
If I got you wrong, please no, because printing kernel memory addresses is never a good idea from the security perspective. user_data is already there

**Comment by axboe** [comment]
We already print memory addresses in a bunch of spots - which should be ok, this is all root only anyway. But in principle I do agree, and if user_data is good enough to match them up, then that might be the best way indeed.

**Comment by isilence** [comment]
I would also remind that in general adding new tracepoints is not free, see fresh gcc output below. So, I wouldn't be so excited adding new ones in hot paths.

```asm
# ./arch/x86/include/asm/atomic.h:29: 	return __READ_ONCE((v)->counter);
movl	__tracepoint_io_uring_complete+8(%rip), %eax
# fs/io_uring.c:2137: 		__io_cqring_fill_event(ctx, req->user_data, req->result,
movq	%r10, %r14	#,
# ./include/linux/jump_label.h:264: 	if (unlikely_notrace(static_key_count(key) > 0))
testl	%eax, %eax	# _51
jg	.L2986	#,
```

**Comment by axboe** [comment]
Right, but I think the point here was more about augmenting existing tracepoints enough to make it possible to discern which events are for which request. And that should definitely be done, it makes them a lot more useful.

**Comment by isilence** [comment]
> I have augmented the io_uring traces to debug my problem. Some of them are inside io_wq_submit_work():
> 
> ```
> 			trace_io_wq_submit_work(req->ctx, req, req->opcode, req->user_data);
> 			ret = io_issue_sqe(req, 0);
> 			trace_io_wq_submit_work_completion(ret);
> ```
> 
> I got a surprising result:
> 
> ```
>  31806.742 iou-wrk-629/637 io_uring:io_wq_submit_work_completion(8)
>  32158.157 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(1)
>  32260.022 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(2)
>  32364.793 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(3)
>  32844.402 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(8)
>  32849.387 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(8)
>  32879.902 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(8)
>  33006.720 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(0)
>  33259.342 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(2)
>  33365.219 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(3)
>  33394.230 iou-mgr-629/638 io_uring:io_wq_submit_work_completion(3)
> ```

It's strange, io_issue_sqe() doesn't return positive results, as well as iou-mgr should not have been submitting I/O itself with an exception when task or io_uring is exiting.
@lano1106, can you confirm those findings with something harder-to-screw like printk? Because looks fancy indeed

**Comment by isilence** [comment]
> Right, but I think the point here was more about augmenting existing tracepoints enough to make it possible to discern which events are for which request. And that should definitely be done, it makes them a lot more useful.

I should have emphasised on __adding new__, agree that modifying current ones is free

**Comment by isilence** [comment]
> We already print memory addresses in a bunch of spots - which should be ok, this is all root only anyway.

I believe the line b/w root/non-root privileged/non-privileged is getting washed out and it's getting trickier.
Checked what it prints -- mainly ctx, but if the idea is just to distinct between different ctxs (it is), maybe better to hash pointers them?

Looks io_uring_fail_link() and some others print req pointer... It's not even consistent, better to get patched up

**Comment by isilence** [comment]
> I would also remind that in general adding new tracepoints is not free, see fresh gcc output below. So, I wouldn't be so excited adding new ones in hot paths.
> 
> ```assembly
> # ./arch/x86/include/asm/atomic.h:29: 	return __READ_ONCE((v)->counter);
> movl	__tracepoint_io_uring_complete+8(%rip), %eax
> # fs/io_uring.c:2137: 		__io_cqring_fill_event(ctx, req->user_data, req->result,
> movq	%r10, %r14	#,
> # ./include/linux/jump_label.h:264: 	if (unlikely_notrace(static_key_count(key) > 0))
> testl	%eax, %eax	# _51
> jg	.L2986	#,
> ```

Btw, it looks strange that it's not dynamically inserted with some magic to make sure all needed vars are accessible. May I have missed something? I think I saw it lighter in asm before.

**Comment by axboe** [comment]
Yeah, I don't care about the exact pointers at all as they aren't useful for what we need, it just need to be unique of course.

**Comment by lano1106** [comment]
> @lano1106, can you confirm those findings with something harder-to-screw like printk? Because looks fancy indeed

I will. There is definitely something wrong with the trace system. I can tell just with the pid. mgr thread is created before the worker one...

**Comment by lano1106** [comment]
> Right, but I think the point here was more about augmenting existing tracepoints enough to make it possible to discern which events are for which request. And that should definitely be done, it makes them a lot more useful.

Yes. Those are the ones that I had in mind to offer. I was thinking of keeping the more experimental ones for my private usage ;-)

The thing is in my io_uring usage, distinct requests can have the same user_data. This is why with the original traces, the root of my problem was unclear. In fact, it was obfuscated by having only the user_data field.

I am not done reading the whole exchange between both of you. I'll make sure of the final verdict before going ahea....

Greetings,

**Comment by lano1106** [comment]
I was about to do more than asked with something along:

```
TRACE_EVENT(io_uring_create,

	TP_PROTO(int fd, void *ctx, u32 sq_entries, u32 cq_entries, u32 flags),

	TP_ARGS(fd, ctx, sq_entries, cq_entries, flags),

	TP_STRUCT__entry (
		__field(  int,		fd			)
		__field(  u32,		ctx_hash	)
		__field(  u32,		sq_entries	)
		__field(  u32,		cq_entries	)
		__field(  u32,		flags		)
	),

	TP_fast_assign(
		__entry->fd			= fd;
		__entry->ctx_hash	= hash_32((u32)ctx, 32);
		__entry->sq_entries	= sq_entries;
		__entry->cq_entries	= cq_entries;
		__entry->flags		= flags;
	),

	TP_printk("ring_hash %u, fd %d sq size %d, cq size %d, flags %d",
			  __entry->ctx_hash, __entry->fd, __entry->sq_entries,
			  __entry->cq_entries, __entry->flags)
);
```

but I will revert to how it was. The naked ptr values is a known security issue by the tracing guys. They already make sure that the ptr values are hashed to display to the regular user. I must have gotten the naked ptr values because I did 'sudo perf' to get the traces.

I am discovering the tracing system. I had no idea it was there before playing with it for my io_uring app debugging.

Here is what I have found:

```
whippet2 /sys/kernel/tracing # cat trace_options 
print-parent
nosym-offset
nosym-addr
noverbose
noraw
nohex
nobin
noblock
trace_printk
annotate
nouserstacktrace
nosym-userobj
noprintk-msg-only
context-info
nolatency-format
record-cmd
norecord-tgid
overwrite
nodisable_on_free
irq-info
markers
noevent-fork
nopause-on-trace
**hash-ptr**
function-trace
nofunction-fork
nodisplay-graph
nostacktrace
notest_nop_accept
notest_nop_refuse
```

I suggest keeping the pointer hashing business centralized to the tracing component...

**Comment by lano1106** [comment]
the patch is almost ready to go. FYI, I did drop io_uring_complete trace modif. The containing function signature did change recently. I prefer to not revert back the signature.

**Comment by lano1106** [comment]
guys you are going to be reliefed. No weird monkey business. The weird output seems to come from trace.

Code for the last run:

```
                        trace_io_wq_submit_work(req->ctx, req, req->opcode, req->user_data);
                        ret = io_issue_sqe(req, 0);
                        printk(KERN_INFO "%s/%d io_wq_submit_work_completion(%d)\n",
                                __get_task_comm(buf, TASK_COMM_LEN, current), current->pid, ret);
                        trace_io_wq_submit_work_completion(ret);
```


trace output:
```
 31943.777 iou-wrk-630/638 io_uring:io_wq_submit_work_completion(9)
 32005.537 iou-wrk-630/639 io_uring:io_wq_submit_work_completion(0)
 32812.644 iou-wrk-630/639 io_uring:io_wq_submit_work_completion(8)
 32863.708 iou-wrk-630/639 io_uring:io_wq_submit_work_completion(8)
 32872.628 iou-wrk-630/639 io_uring:io_wq_submit_work_completion(8)
 33066.644 iou-wrk-630/639 io_uring:io_wq_submit_work_completion(0)
 33421.610 iou-wrk-630/639 io_uring:io_wq_submit_work_completion(4)
```

printk output:

```
[  733.933898] iou-wrk-630/638 io_wq_submit_work_completion(0)
[  733.995658] iou-wrk-630/639 io_wq_submit_work_completion(0)
[  734.802765] iou-wrk-630/639 io_wq_submit_work_completion(0)
[  734.853828] iou-wrk-630/639 io_wq_submit_work_completion(0)
[  734.862748] iou-wrk-630/639 io_wq_submit_work_completion(0)
[  735.056766] iou-wrk-630/639 io_wq_submit_work_completion(0)
[  735.411731] iou-wrk-630/639 io_wq_submit_work_completion(0)
[  735.850804] iou-wrk-630/636 io_wq_submit_work_completion(0)
```
If you know the trace maintainer, maybe make him take a look at this experiment here...

**Comment by lano1106** [comment]
BTW, my bug making some io-worker threads blocked doing some i/o may be in part responsible for my core dump issue at #346

Of all the things that I have tried to capture in my test program to replicate the core dump issue, the buggy behavior of sending more than one async read requests for the same socket at the same time is something that I did not attempt...

**Comment by isilence** [comment]
> BTW, my bug making some io-worker threads blocked doing some i/o may be in part responsible for my core dump issue at #346

No, it's expected them doing that

**Comment by isilence** [comment]
> guys you are going to be reliefed. No weird monkey business. The weird output seems to come from trace.

No worries, it was expected

> If you know the trace maintainer, maybe make him take a look at this experiment here...

I'd actually first double check the tracing you added, because it's more likely the cause.

**Comment by lano1106** [comment]
> > If you know the trace maintainer, maybe make him take a look at this experiment here...
> 
> I'd actually first double check the tracing you added, because it's more likely the cause.

then you will need its definition:

```
TRACE_EVENT(io_wq_submit_work_completion,

        TP_PROTO(int ret),

        TP_ARGS(ret),

        TP_STRUCT__entry (
                __field(  int,     ret      )
        ),

        TP_fast_assign(
                __entry->ret       = ret;
        ),

        TP_printk("ret %d",
                          __entry->ret)
);
```

**Comment by lano1106** [comment]
> > BTW, my bug making some io-worker threads blocked doing some i/o may be in part responsible for my core dump issue at #346
> 
> No, it's expected them doing that

yes it is supposed to block but it should eventually get completed and return back to io_wqe_worker(). In my case, I believe that they may indefinitely remain blocked and the core dump generation is happening while the thread are blocked there.

It is a bit unusual how they handle signals. Unless we craft a test to verify that all is well when a core dump is created exactly when they are blocked, I suspect that this might be a rare occurence happening naturally.

Anyway, I propose to double check to make sure that all is good in that scenario.

**Comment by isilence** [comment]
Do you see worker tasks left after your app crashed and returned?

**Comment by isilence** [comment]
> no

Good. Workers can block for long until they get some data, and it may happen even with poll stuff preceding it.

For the verification part, just send a pipe read with IOSQE_ASYNC with nobody writing to it (so it will block a worker), or even a bunch of such. And then crash as usual for coredumping

**Comment by isilence** [comment]
> no. I guess that it would normally mean that all went well... but I still have a doubt. My program is in a specific state that is stopping it generate the core. I have been struggling to recreate the conditions for that happening and I haven't been able to find out what. As of today, this lead is the only and best that I have to explain the core dump problem...

If I would guess, it may be signal handling. Not familiar with that part, but should be
`fs/coredump.c do_coredump()`, called from `kernel/signal.c get_signal()`. You may debug it and see whether it gets into the do_coredump() and whether it even gets there and why (tracing get_signal()).

**Comment by lano1106** [comment]
> > no. I guess that it would normally mean that all went well... but I still have a doubt. My program is in a specific state that is stopping it generate the core. I have been struggling to recreate the conditions for that happening and I haven't been able to find out what. As of today, this lead is the only and best that I have to explain the core dump problem...
> 
> If I would guess, it may be signal handling. Not familiar with that part, but should be
> `fs/coredump.c do_coredump()`, called from `kernel/signal.c get_signal()`. You may debug it and see whether it gets into the do_coredump() and whether it even gets there and why (tracing get_signal()).

it makes sense to me too. It does reach the do_coredump() part since I wouldn't get the empty size or truncated core file otherwise. If something wrong happens at that point, it must be around the part where all threads of the group get zapped.

This is on my todo list to review that code and make sense of it in the particular io_uring context.

**Comment by isilence** [comment]
E.g. do worker threads go into do_coredump(), and if so shouldn't we rather ignore it and move PF_IO_WOKER check before sig_kernel_coredump(). Or see if signals are somehow shared and workers do hijack it somehow.

All is a speculation though without taking a good look

**Comment by lano1106** [comment]
I know what you mean. The issue resolve around how deliver_signal is picking the thread to handle the signal and how the io-wrk threads react to the event based on their particular signal settings. Not sure yet about the details but my gut feeling is telling me that the explanation is somewhere in that spot.

I just made the pipe test. I got a nice core dump so this is not it... I still need search elsewhere...
